### PR TITLE
fix: [CDS-48334]: Save workloads and releaseHistory even if apply fails

### PR DESCRIPTION
**Pull request by vikyathharekal**
## Describe your changes
We are not storing the managed workloads and release history in case the apply step fails in a Rolling Deployment.
In this customer scenario, we have a Canary Deployment that is successful in the same workflow and we don’t store managed workloads in the release history in this case by design. Hence we assume that there are no managed workloads at all.
The fix is to store the managed workloads and release history even when apply step fails. Both in CG and NG.

## Testing (CG & NG)
1. Rolling 
2. Rolling Rollback when it fails before apply step
3. Rolling Rollback when it fails during apply step
4. Canary with Rolling
5. Canary with Rolling Rollback when it fails before apply step
6. Canary with Rolling Rollback when it fails during apply step
7. When manifest is changed to have workload x, instead of y and apply fails.
8. With pruning enabled.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41512)

### CODE REVIEW

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sRollingRequestHandler.java [hunk 1]
```diff
@@ -133,17 +133,21 @@ protected K8sDeployResponse executeTaskInternal(K8sDeployRequest k8sDeployReques
     List<K8sPod> existingPodList = k8sRollingBaseHandler.getExistingPods(
         steadyStateTimeoutInMillis, allWorkloads, kubernetesConfig, releaseName, prepareLogCallback);
 
-    k8sTaskHelperBase.applyManifests(client, resources, k8sDelegateTaskParams,
-        k8sTaskHelperBase.getLogCallback(logStreamingTaskClient, Apply, true, commandUnitsProgress), true, true);
-    shouldSaveReleaseHistory = true;
+    try {
+      k8sTaskHelperBase.applyManifests(client, resources, k8sDelegateTaskParams,
+          k8sTaskHelperBase.getLogCallback(logStreamingTaskClient, Apply, true, commandUnitsProgress), true, true);
+      shouldSaveReleaseHistory = true;
```
Comment (vikyathharekal) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sRollingRequestHandler.java:139]: do this before applyManifests

Comment (vikyathharekal) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/K8sRollingRequestHandler.java:139]: how does kubectl behave when we undo to same version?

**Review by GoelPratyush99 — APPROVED** [review]
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41512)* status: 4 file(s) to review. 1 of 2 approvals obtained.

**Review by vikyathharekal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41512)* status: 1 file(s) to review. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/test/java/io/harness/delegate/k8s/K8sRollingRequestHandlerTest.java` line 132 at r3](https://reviewable.io/reviews/harness/harness-core/41512#-NJFiT0_2KoitGM2M4Lg:-NJFlYRq1UCT33mNtCmy:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/dc3619c9ea0614968a9865bd9d1d2a2db0d2699c/930-delegate-tasks/src/test/java/io/harness/delegate/k8s/K8sRollingRequestHandlerTest.java#L132)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

nit: wrong owner
</blockquote></details>

Done

**Review by abosii-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41512)* status: 1 file(s) to review. 1 of 2 approvals obtained.

**Review by GoelPratyush99 — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41512)* status: 1 file(s) to review.

**Review by GoelPratyush99 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41512)* status: 1 file(s) to review.

**Review by GoelPratyush99 — APPROVED** [review]
> Reviewed 1 of 2 files at r1, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41512)* status: :shipit: complete! ✓

**Pull request merged**